File: android/src/main/java/io/parity/signer/components/networkicon/jdenticon/jdenticon_kotlin/jdenticon.kt
package io.parity.signer.components.networkicon.jdenticon.jdenticon_kotlin

internal class Jdenticon {

    companion object {

        /**
         * Draws an identicon as an SVG string.
         * @param hashOrValue - A hexadecimal hash string or any value that will be hashed by Jdenticon.
         * @param size - Icon size in pixels.
         * @param padding - Optional padding in percents. Extra padding might be added to center the rendered identicon.
         * @returns SVG string
         */
        fun toSvg(hashOrValue: String, size: Int, padding: Float? = null): String {
            val hash = HashUtils.keepOrCreateHash(hashOrValue)
            var writer = SvgWriter(size)
            var renderer = SvgRenderer(writer)
            IconGenerator(
                    renderer,
                    hash,
                    0f,
                    0f,
                    size.toFloat(),
                    padding,
                    getCurrentConfig()
            );
            return writer.toString()
        }

        /**
         * Gets the normalized current Jdenticon color configuration. Missing fields have default values.
         */
        fun getCurrentConfig(): Config {
            var backColor = "#FFFFFF"
            /**
             * Creates a lightness range.
             */
            fun lightness(configName: String, defaultMin: Float, defaultMax: Float): (Float) -> Float {
                var range = arrayOf(defaultMin, defaultMax)

                /**
                 * Gets a lightness relative the specified value in the specified lightness range.
                 */
                return fun(value: Float): Float {
                    var value2 = range[0] + value * (range[1] - range[0])
                    return if (value2 < 0) 0f else if (value2 > 1) 1f else value2
                }
            }

            return Config(
                    0.5f,
                    lightness("color", 0.4f, 0.8f),
                    lightness("grayscale", 0.3f, 0.9f),
                    Color.parse(backColor)
            )
        }
    }
}

File: ios/Packages/Jdenticon/Sources/Jdenticon/Public/Jdenticon.swift
//
//  Jdenticon.swift
//
//
//  Created by Krzysztof Rodak on 23/08/2023.
//

import CoreGraphics
import UIKit

final class Jdenticon {
    private let hashGenerator: HashGenerator

    init(hashGenerator: HashGenerator = HashGenerator()) {
        self.hashGenerator = hashGenerator
    }

    func toSvg(hashOrValue: String, size: Int, padding: CGFloat? = nil) -> String {
        let hash = hashGenerator.keepOrCreateHash(hashOrValue)
        let writer = SvgWriter(size: size)
        let renderer = SvgRenderer(target: writer)
        _ = IconGenerator(
            renderer: renderer,
            hash: hash,
            x: 0,
            y: 0,
            size: CGFloat(size),
            padding: padding,
            config: getCurrentConfig()
        )
        return writer.description
    }

    func getCurrentConfig() -> Config {
        let backColor = "#FFFFFF"

        func lightness(defaultMin: CGFloat, defaultMax: CGFloat) -> (CGFloat) -> CGFloat {
            let range: [CGFloat] = [defaultMin, defaultMax]

            return { value in
                let value2 = range[0] + value * (range[1] - range[0])
                if value2 < 0 {
                    return 0
                } else if value2 > 1 {
                    return 1
                } else {
                    return value2
                }
            }
        }

        return Config(
            saturation: 0.5,
            colorLightness: lightness(defaultMin: 0.4, defaultMax: 0.8),
            grayscaleLightness: lightness(defaultMin: 0.3, defaultMax: 0.9),
            backColor: Color.parse(backColor)
        )
    }
}
